C/Adrian Ashley/FNORDADEL/man/mshrink.1
.TH MSHRINK 1 foo bar "FNORDADEL BBS SOFTWARE"
.SH NAME
mshrink.tos - shrink the the Fnordadel message file
.SH SYNOPSIS
mshrink [pathname]
.SH DESCRIPTION
Mshrink will determine the point (in kilobytes) in your message
file, ctdlmsg.sys, at which the next message will be saved.  The
program will give you the option of reducing the size of the
ctdlmsg.sys to that point.  If you answer "yes", it will cut the
message base, prompt you to change your ctdlcnfg.sys file as
required, and to run configur.tos.  If you do not do these two
things, the fleas of a thousand camels will infest your storage
devices.  I.e., it will be a Bad Thing(tm).
.PP
Mshrink expects to be run from the directory containing your
ctdltabl.sys file.  
.SH OPTIONS
.IP [pathname]
The full pathname of a directory in which to create the
temporary message file during compaction.  See NOTES.
Examples: "c:\\tmp", "a:\\", "e:\\foo\\bar\\baz"
.SH NOTES
Due to the way all Citadels organise their messages, it is not
feasible to shrink the message base to any arbitrary size.  Thus
if you desire some particular size, you will need to run this
program regularly and wait until the current message pointer
gets close to the size you want.  Then give mshrink free rein to
do its worst.
.PP
During the shrinkage, mshrink will create a temporary file of
the new, smaller size.  You must, therefore, have enough room
on your drive(s) to hold both the old and* new message files during
the compaction process.  If you don't have room, as you might not
if you run on floppy drive(s), you may use the optional command
line parameter to specify the full pathname of a directory in which
to put the temporary file (e.g. another floppy drive or a RAM disk).
.PP	
BACKUP your system!  If you run this program without first backing
up your system, the Imp of the Perverse may well murder your system
and dance on its grave, singing "Hallelujah!"  And it won't be our
fault.
.PP
When mshrink gives your message file the chop, you will lose
old messages on the system which inhabited the region of the file
that was purged.
.SH FILES
.DT
.ta \w'#msgdir\\ctdlmsg.sys\ \ \ 'u
.br
mshrink.tos	The message file compactor.
.br
#msgdir\\ctdlmsg.sys	The message file.
.br
ctdlcnfg.sys	The configuration file.
.br
ctdltabl.sys	The system tables file.
.br
.SH SEE ALSO
The Fnordadel Reference Manual
.SH AUTHORS
Written by Adrian Ashley and Royce Howland for the Fnordadel
project.


C/Adrian Ashley/FNORDADEL/citutil/mshrink.c
/*
 * mshrink.c -- Shrink the Fnordadel message file, on approval.
 *
 * 90May29 A&R	Created as part of 1.3 release.
 */

#include "ctdl.h"
#include "msg.h"
#include "config.h"
#include "citlib.h"

void doCR() {}
mAbort() {}
mPrintf() {}
xprintf() {}	/* because something somewhere is linking in libroute.o */

char prevChar;
char echo;
#if 0
struct aRoom roomBuf;		/* to avoid linking in libroom.o	*/
#endif

char *program = "mshrink";
int tempdir = NO, tempfl;
PATHBUF tempfile, msgfile;
char buf[BLKSIZE];

/* Get a yes/no response from console */
int
getyesno(void)
{
    int c;

    while (1) {
	c = toupper(getch());
	if (c == 'Y' || c == 'N')
	    break;
    }
    putchar(c);
    putchar('\n');
    if (c == 'N')
	return NO;
    else
	return YES;
}

/* Copy 'num' blocks of size BLKSIZE from mfl/mfile to tfl/tfile */
void
mcopy(int mfl, char *mfile, int tfl, char *tfile, int num)
{
    register int i;

    for (i = 1; i <= num; i++) {
	printf("\rCopying block %d of %d", i, num);
	fflush(stdout);
	if (dread(mfl, &buf[0], BLKSIZE) != BLKSIZE)
	    crashout("could not read %d bytes from block %d of %s", BLKSIZE,
		i, mfile);
#if 0
	if (i == num)	/* Nuke remnants of last sector */
	    memset(&buf[cfg.catChar], 0, BLKSIZE - cfg.catChar);
#endif
	if (dwrite(tfl, &buf[0], BLKSIZE) != BLKSIZE)
	    crashout("could not write %d bytes to block %d of %s", BLKSIZE,
		i, tfile);
    }
    printf("\rCopied %d blocks from %s to %s.\n", num, mfile, tfile);
}

/* Chop the message file to 'size' K */
void
chopit(int size)
{
    long msg;

    ctdlfile(msgfile, cfg.msgdir, "ctdlmsg.sys");
    if ((msgfl = dopen(msgfile, O_RDONLY)) < 0)
	crashout("cannot open %s", msgfile);
    if ((tempfl = dcreat(tempfile)) < 0)
	crashout("cannot create %s", tempfile);

    msgseek(0, 0);	/* Find ID# of message in ctdlmsg.sys sector 0 */
    getmessage();
    msg = msgBuf.mbid;
    printf("The oldest message used to be %ld; it's now %ld.\n", cfg.oldest,
	msg);
    printf("The message file size used to be %dK; it's now %dK.\n",
	(int) ((long)cfg.maxMSector * (long)BLKSIZE / 1024L), size);

    cfg.oldest = msg;
    cfg.maxMSector = cfg.catSector + 1;

    dseek(msgfl, 0L, 0);	/* Return file pointer to start of file */
    mcopy(msgfl, msgfile, tempfl, tempfile, cfg.maxMSector);
    dclose(msgfl);
    dclose(tempfl);
    if (dunlink(msgfile))	/* Kill old ctdlmsg.sys */
	crashout("cannot unlink %s", msgfile);

    if (tempdir) {		/* Did we use a separate temp directory? */
/* next line changed to tempfl from msgfl by AA 90Oct04 */
	if ((tempfl = dopen(tempfile, O_RDONLY)) < 0)
	    crashout("cannot open %s", tempfile);
	if ((msgfl = dcreat(msgfile)) < 0)
	    crashout("cannot create new %s", msgfile);
	mcopy(tempfl, tempfile, msgfl, msgfile, cfg.maxMSector);
	dclose(tempfl);
	dclose(msgfl);
	if (dunlink(tempfile))		/* Kill temp file */
	    crashout("cannot unlink %s", tempfile);
    }
    else	/* Used #sysDir, so just rename temp file to ctdlmsg.sys */
	if (drename(tempfile, msgfile))
	    crashout("cannot rename %s to %s", tempfile, msgfile);
}

main(int argc, char **argv)
{
    int p, pos, junk;

    setbuf(stdout, NULL);

    printf("%s for Fnordadel V%s\n", program, VERSION);

    tempfile[0] = 0;
    if (argc == 2) {
	tempdir = YES;		/* Flag that we're using another directory */
	strcat(tempfile, argv[1]);
	junk = strlen(tempfile);
	if (tempfile[junk - 1] == '\\')
	    tempfile[junk - 1] = 0;
	strcat(tempfile, "\\ctdlmsg.tmp");
	printf("\nThe temporary file will be %s.\n", tempfile);
	printf("If this looks wrong, don't let mshrink do the chop.\n");
    }
    else if (argc > 2) {
	fprintf(stderr, "usage: %s [pathname]\n", program);
	if (fromdesk())
	    hitkey();
	exit(1);
    }

    if (readSysTab(FALSE) && makelock(&p)) {
	if (!tempfile[0])
	    ctdlfile(tempfile, cfg.msgdir, "ctdlmsg.tmp");
	pos = (int) ((long)(cfg.catSector + 1) * (long)BLKSIZE / 1024L);
	printf("\nThe message pointer is currently %dK into the message file.\n",
	    pos);
	printf("Do you wish to chop the message file at this size? (y/n) ");
	if (getyesno()) {
	    putchar('\n');
	    if (cfg.catSector == cfg.maxMSector - 1)
		printf("\nThe message file is already that size.\n");
	    else {
		chopit(pos);
		printf("\nBe sure to change ctdlcnfg.sys to reflect the new value of\n");
		printf("messagek (%d).  Then run configur.  If you don't, your\n",
		    pos);
		printf("system will explode.\n"); 
		writeSysTab();
	    }
	}
	wipelock(&p);
    }
    if (fromdesk())
	hitkey();
    exit(0);
}
